---
title: "count_5SrDNA"
output: html_document
---

# PROCESSING OF GFF FILES

```{r}
workdir <- '/GitHub/Ey15-2_vs_Col-0/'

setwd(workdir)


library(ggplot2)
library(gridExtra)
library(ggthemes)



# Load accession names from universal list
files <- list.files(paste(workdir, "/gff_files/", sep = ""), pattern = "*5S_rDNA.gff")
files
accs <- unlist(lapply(strsplit(files, "[.]"), "[", 1))
accs


```

```{r}


gff <- data.frame(Chromosome=NA, Type=NA, Start=NA , End=NA, Orientation=NA, Length=NA, Divergence=NA, Accession=NA)

  
for(i in 1:length(accs)){
  # Read both rDNA and rRNA
  gff_rDNA <- read.table(paste(workdir, '/gff_files/', accs[i], '.5S_rDNA.gff', sep = ''))
  gff_rRNA <- read.table(paste(workdir, '/gff_files/', accs[i], '.5S_rRNA.gff', sep = ''))

  # Concatenate them
  gff_tmp <- rbind(gff_rDNA, gff_rRNA)
  
  #Guarantee only chromosomes
  gff_tmp <- gff_tmp[ grep("Chr", gff_tmp$V1), ]
  
  #Get rid of ugly suffix in Diff lines names
  gff_tmp$V1 <- gsub(pattern = "_RagTag_RagTag", replacement = "", gff_tmp$V1)
  
  # Split useful info columns
  gff_tmp$Length <- unlist(lapply(strsplit(as.character(gff_tmp$V9), "Length="), "[", 2))
  gff_tmp$Length <- as.numeric(unlist(lapply(strsplit(as.character(gff_tmp$Length), "[;]"), "[", 1)))

  gff_tmp$Divergence <- unlist(lapply(strsplit(as.character(gff_tmp$V9), "Divergence="), "[", 2))
  gff_tmp$Divergence <- as.numeric(unlist(lapply(strsplit(as.character(gff_tmp$Divergence), "[;]"), "[", 1)))

  # Get rid of unnecessary columns
  gff_tmp <- gff_tmp[ , -c(2,6,8,9)]
  colnames(gff_tmp)[1:5] <- c('Chromosome', 'Type', 'Start', 'End', 'Orientation')

  # Add accession
  gff_tmp$Accession <- accs[i]

  # Add to final data frame
  gff <- rbind(gff, gff_tmp)
  
}  
  
gff <- gff[-1,]

head(gff)

rDNA <- subset(gff, Type == "5S_rDNA")

# Remove entries based on divergence and 
dim(rDNA)
#[1] 12852     8

# Keep only 5S rDNA repeats larger than 200bp and with a divergence lower than 15
rDNA_filt <- subset(rDNA, Length > 200 & Divergence < 15)
dim(rDNA_filt)
#[1] 12406     8

rDNA_filt


```

# OVERLAPS INTO BLOCKS

```{r}
library(data.table)

# Define chromosomes
chromosomes <- data.frame(Chr=c('Chr1', 'Chr2', 'Chr3', 'Chr4', 'Chr5'), Size=c(30427671, 19698289, 23459830, 18585056, 26975502), stringsAsFactors=FALSE)

# Define signs
signs <- c("+", "-")


# Declare object container
rDNA_blocks <- data.frame(Chromosome=NA, Repeat=NA, Start=NA, End=NA, Orientation=NA, Length=NA, Accession=NA)

# Loop accession
for(a in 1:length(accs)){
# Loop Chromosome
  for(c in 1:dim(chromosomes)[1]){
# Loop Orientation
    for(o in 1:length(signs)){
  
      # subset single group to be analyzed            
      rDNA_ranges <- subset(rDNA_filt, Accession == accs[a] & Chromosome == chromosomes[c,1] & Orientation == signs[o])
      rDNA_ranges <- rDNA_ranges[,c(3,4)]
      
      # Condition to check if object has rows
      if(dim(rDNA_ranges)[1] > 0){
        # Overlapping
        setDT(rDNA_ranges)
        rDNA_ranges <-  rDNA_ranges[, .(start=min(Start), stop=max(End)),
         by=.(group=cumsum(c(1, tail(Start-3, -1) > head(End+3, -1))))]

        # Bring all meta information for the block
        rDNA_blocks_TMP <- data.frame(Chromosome=chromosomes[c,1], Repeat="5S_rDNA", Start=rDNA_ranges$start, End=rDNA_ranges$stop, Orientation=signs[o], Length=rDNA_ranges$stop-rDNA_ranges$start, Accession=accs[a])
      
        # Merge with larger dataset
        rDNA_blocks <- rbind(rDNA_blocks, rDNA_blocks_TMP)
      }
    }
  }
}
# Remove first empty row
rDNA_blocks <- rDNA_blocks[-1,]

# Sort by coordinate
rDNA_blocks <- rDNA_blocks[  with(rDNA_blocks, order(Accession, Chromosome, Start)), ]

rDNA_blocks

```









```{r}


```

# Col-0 genomes

```{r}
# Blocks for Benchmarking paper, comparison with Col-0

rDNA_blocks_Col <- subset(rDNA_blocks, Accession == "Col-0" | Accession == "Naish2021" | Accession == "Wang2021" | Accession == 9994 )

# Change names
rDNA_blocks_Col$Accession <- sub("9994", "9994 This study", rDNA_blocks_Col$Accession)
rDNA_blocks_Col$Accession <- sub("Col-0", "6909 This study", rDNA_blocks_Col$Accession)
rDNA_blocks_Col$Accession <- sub("Wang2021", "Wang et al. 2021", rDNA_blocks_Col$Accession)
rDNA_blocks_Col$Accession <- sub("Naish2021", "Naish et al. 2021", rDNA_blocks_Col$Accession)

rDNA_blocks_Col

rDNA_acc_chr_Col <- aggregate(rDNA_blocks_Col$Length, by=list(rDNA_blocks_Col$Chromosome, rDNA_blocks_Col$Accession, rDNA_blocks_Col$Orientation), sum)
colnames(rDNA_acc_chr_Col) <- c('Chromosome', 'Accession', 'Orientation', 'Length')

rDNA_acc_chr_Col



```

- Almost no rDNA in Chr for 9994

Barplot, comparison of ALL Col-0 and 9994

```{r}

rDNA_acc_chr_Col$Type <- sapply(rDNA_acc_chr_Col$Accession, FUN=function(x){
  if(x=='9994 This study'){
    x <- 'Ey15-2 (id: 9994)'
  }else{
    x <- 'Col-0    (id: 6909)'
  }
})
rDNA_acc_chr_Col$Orientation <- sub("-", "\u2013", rDNA_acc_chr_Col$Orientation, fixed=TRUE)


rDNA_acc_chr_Col$Type <- paste(rDNA_acc_chr_Col$Orientation, rDNA_acc_chr_Col$Type, sep="   ")


# Sort levels TYPE
rDNA_acc_chr_Col$Type <- factor(rDNA_acc_chr_Col$Type, levels = c("\u2013   Ey15-2 (id: 9994)", "+   Ey15-2 (id: 9994)", "\u2013   Col-0    (id: 6909)", "+   Col-0    (id: 6909)"))
levels(factor(rDNA_acc_chr_Col$Type))

# Sort levels ACCESSION
rDNA_acc_chr_Col$Accession <- factor(rDNA_acc_chr_Col$Accession, levels = c("9994", "9994 This study", "Col-0", "6909 This study", "MA31-59", "MA line   59 (gen. 32)", "MA31-69", "MA line   69 (gen. 32)", "MA31-89", "MA line   89 (gen. 32)", "MA31-119", "MA line 119 (gen. 32)", "Wang2021", "Wang et al. 2021", "Naish2021", "Naish et al. 2021" ))
levels(factor(rDNA_acc_chr_Col$Accession))


# In the case of 5S, remove Chr1 and Chr2
#rDNA_acc_chr_Col <- subset(rDNA_acc_chr_Col, Chromosome != "Chr1" & Chromosome != "Chr2" )

#png(paste(workdir, "/plots/5SrDNA_comparison.9994_Col-0.png", sep = ""), width = 2800, height = 1800, res = 400)
ggplot(rDNA_acc_chr_Col, aes(x = Accession, y = Length/1000000, fill = Type)) +
    geom_col(position = "stack", stat = "identity") + 
    scale_fill_manual(values=c("#dfc27d", "#b6843f",  "#80cdc1", "#08979c", "#80cdc1", "#08979c", "#80cdc1", "#08979c", "#80cdc1", "#08979c", "#80cdc1", "#08979c")) +
    facet_grid(. ~ Chromosome) +
    theme_bw() +
    xlab("") + ylab("5S rDNA length (Mb)") +
    ggtitle("5S rDNA") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 
#dev.off()


```
